Utilities/AzureOnboardingCLI/azure-automatic-role-assignment.py

```diff
@@ -461,64 +501,98 @@ def main():
+                        products,
+                        args.appRegistrationId
+                    )
                 )
-            )
-            if account_id is not None and token is not None:
-                set_azure_credentials(
-                    token,
-                    account_id,
-                    required_parameters_response["appId"],
-                    required_parameters_response["password"],
-                    required_parameters_response["tenant"],
-                    subscription_id,
+                if account_id is not None and token is not None:
+                    set_azure_credentials(
+                        token,
+                        account_id,
+                        required_parameters_response["appId"],
+                        required_parameters_response["password"],
+                        required_parameters_response["tenant"],
+                        subscription_id,
+                    )
+
+                register_products(account_id, token, products)
+                display_result(subscription_id, required_parameters_response)
+
+                log(f"Completed onboarding subscription {index+1} of {len(subscription_ids)} ({subscription_name} - {subscription_id})\n")
+                number_successfully_onboarded += 1
+
+            except SpotinstClientException as e:
+                # the spotinst token is invalid. halt all onboarding
+                log(f"Spot client exception.  Error: {str(e)}")
+                number_failed_onboarded += 1
+                break
+            except ConnectTimeout as e:
+                # the Spot API cannot be reached.  halt all onboarding
+                log(f"Spot client connection has timed out.  Error: {str(e)}")
+                number_failed_onboarded += 1
+                break
+            except Exception as e:
+                '''
+                a non-transaction breaking exception occurred. stop processing this subscription but continue if
+                there are more subscriptions to be processed
+                '''
+                log(
+                    f"Error occurred during onboarding process for subscription {subscription_name} - {subscription_id}. Reason: {str(e)}",
+                    LogLevel.ERROR,
                 )
-
-            register_products(account_id, token, products)
-            print(f"Completed onboarding subscription {subscription_name} ({subscription_id})")
-            display_result(subscription_id, required_parameters_response)
-
-        print("Finished creating required credentials parameters for Spot registration")
+                number_failed_onboarded += 1
+                continue
+
+        # print the summary here
+        log("Operation completed.  Summary:")
+        log(f"    Number of subscriptions attempted:                {len(subscription_ids)}")
+        log(f"    Number of subscriptions onboarded successfully:   {number_successfully_onboarded}")
```

Comment:
Can we dump the subscription IDs that failed?

Reply:
Updated